### PR TITLE
[FEAT] 서술형 제출 채점 로직 구현

### LINKED ISSUE #81: [FEAT] 서술형 제출 채점 로직 구현
### 기능 요약 (Summary)

123

### 상세 내용 (Detail)

123

### 참고 자료 (References)

123

### PR DESCRIPTION
**Pull request by PSW99**
## 📝 개요 (Description)

  QuizService에 서술형(essay) 문제 채점 기능을 추가했습니다. 기존에는 객관식과 단답형만 지원했지만, 이제 서술형 문제도 AI를 통해 0~10점 사이의 부분 점수를 부여할 수 있습니다.

  **주요 변경사항:**
  - 서술형 문제 채점 시 부분 점수(0~10점) 지원
  - 7점 이상은 정답(isCorrect: true), 7점 미만은 오답으로 처리
  - 채점 프롬프트에 문제 타입별 상세 채점 기준 추가
  - 서술형 문제 채점 로직 및 테스트 코드 작성

  ## 🔗 관련 이슈 (Related Issues)

  - Closes #81

  ## 🏷️ 작업 유형 (Type of Change)

  - [x] ✨ 기능 추가 (New Feature)
  - [ ] 🐛 버그 수정 (Bug Fix)
  - [ ] ♻️ 리팩토링 (Refactoring)
  - [ ] 📝 문서 업데이트 (Documentation)
  - [ ] 🔧 설정 변경 및 기타 (Config & Other)

  ## 📋 변경 내역 (Changes)

  ### 1. `quiz-prompts.ts`
  - GRADER 프롬프트 업데이트
  - 문제 타입별(객관식/단답형/서술형) 채점 기준 명시
  - 서술형 부분 점수 기준 추가:
    - 10점: 모범 답안과 거의 동일, 핵심 개념 모두 정확히 설명
    - 7-9점: 핵심 개념 대부분 이해, 일부 누락/부정확
    - 4-6점: 일부 핵심 개념만 언급, 이해 부족
    - 1-3점: 관련 내용 일부 있으나 핵심 놓침
    - 0점: 완전히 틀림/무관한 답변

  ### 2. `quiz.service.ts`
  - `gradeRound` 메서드 수정
    - 문제 타입을 userMessage에 포함
    - 서술형 문제의 경우 score description을 "0~10점 사이의 부분 점수"로 설정
    - 객관식/단답형은 기존처럼 10점 또는 0점

  ### 3. `quiz.service.spec.ts`
  - `gradeRound` 메서드에 대한 포괄적인 테스트 추가 (7개 테스트 케이스)
    - 객관식 문제 채점 (정답 10점, 오답 0점)
    - 단답형 문제 채점 (정답 10점, 오답 0점)
    - 서술형 문제 부분 점수 (9점, 3점, 1점)
    - 서술형 경계값 테스트 (7점 = 정답 처리)
    - 문제 타입 포함 확인
    - essay 타입에서 sampleAnswer 사용 확인

  ## ✅ 체크리스트 (Checklist)

  - [x] 코드가 정상적으로 컴파일/빌드 되나요?
  - [x] 기존 테스트를 통과했나요? (새로운 테스트가 필요하다면 추가했나요?)
  - [x] 스스로 코드를 리뷰했나요? (Self-review)
  - [x] 불필요한 주석이나 디버깅 코드는 제거했나요?
  - [] 문서(README 등)에 변경 사항을 업데이트했나요?

  ## 🧪 테스트 결과 (Test Results)

  ```bash
  PASS test/quiz/quiz.service.spec.ts
    QuizService
      generateQuestion
        ✓ should return 5 questions from DB with balanced difficulty and type
        ✓ should convert difficulty correctly
        ✓ should handle short answer questions
        ✓ should handle essay questions
        ✓ should handle null difficulty as medium
        ✓ should throw error when less than 5 questions available
        ✓ should throw error when JSON parsing fails for multiple choice
        ✓ should throw error when correctAnswer is missing for short answer
      gradeRound
        ✓ should grade multiple choice question with correct answer
        ✓ should grade short answer question with correct answer
        ✓ should grade essay question with partial scores
        ✓ should handle essay question with score exactly 7 (boundary case)
        ✓ should include question type in user message
        ✓ should use sampleAnswer for essay questions in user message

  Test Suites: 3 passed, 3 total
  Tests:       18 passed, 18 total
```

  ## 🎸 기타 (Etc)

  리뷰 포인트:
  1. 서술형 채점 기준(0~10점, 7점 이상 정답)이 적절한지 검토 부탁드립니다
  2. AI 프롬프트의 채점 기준이 명확한지 확인 부탁드립니다
  3. 테스트 커버리지가 충분한지 의견 주시면 감사하겠습니다

  추가 고려사항:
  - 향후 서술형 문제의 키워드 기반 추가 검증 로직 도입 가능
  - 채점 결과에 대한 이의 제기 시스템 구축 필요 여부 검토

### CODE REVIEW

**Review by MINU234 — APPROVED** [review]
고생하셨습니다 ~!

File: packages/backend/src/quiz/quiz-prompts.ts [hunk 1]
```diff
@@ -1,20 +1,27 @@
...
+   - **isCorrect:** 7점 이상이면 true, 7점 미만이면 false.
+   - **score:** 0~10점 사이의 부분 점수를 부여해. 다음 기준을 참고:
+     * 10점: 모범 답안과 거의 동일하거나 핵심 개념을 모두 정확히 설명
+     * 7-9점: 핵심 개념을 대부분 이해하고 설명했으나 일부 누락 또는 부정확함
+     * 4-6점: 일부 핵심 개념만 언급했거나 이해가 부족함
+     * 1-3점: 관련 내용이 일부 있으나 핵심을 놓침
+     * 0점: 완전히 틀렸거나 무관한 답변
+   - **feedback:**
+     - 7점 이상: 잘한 점을 구체적으로 칭찬하고, 더 보완하면 좋을 점을 간략히 언급 (2-3문장)
+     - 7점 미만: 놓친 핵심 개념이나 오류를 구체적으로 지적하고, 올바른 답변의 방향을 제시 (2-3문장)
```
Comment (NAKTA-Y) [inline packages/backend/src/quiz/quiz-prompts.ts]: 저희 AI 피드백 기능은 should 로 분류 되어있고 우선순위가 낮았던걸로 기억합니다!

**Review by NAKTA-Y — COMMENTED** [review]
수고하셨습니다

File: packages/backend/src/quiz/quiz.service.ts [hunk 2]
```diff
@@ -342,25 +335,26 @@ export class QuizService {
...
     });
 
-    // 결과 매핑 (원본 답안 텍스트 복원 및 초기 점수 0점 세팅)
+    // 결과 매핑 (원본 답안 텍스트 복원)
     return result.grades.map((grade) => {
       const originalSubmission = submissions.find((s) => s.playerId === grade.playerId);
 
       return {
         playerId: grade.playerId,
         answer: originalSubmission ? originalSubmission.answer : '',
```
Comment (Enble) [inline packages/backend/src/quiz/quiz.service.ts]: ai가 submission에 없는 playerId를 반환하거나, playerId를 누락하는 경우에 대한 검증이 필요합니다

File: packages/backend/src/quiz/quiz-prompts.ts (hunk 1 above)
Comment (NAKTA-Y) [reply packages/backend/src/quiz/quiz-prompts.ts]: 기능에 변화가 크지 않다면 남겨두셔도 괜찮을 것 같아요!

File: packages/backend/src/quiz/quiz.service.ts [hunk 3]
```diff
@@ -342,25 +335,26 @@ export class QuizService {
...
         isCorrect: grade.isCorrect,
-        score: 0, // 점수는 GameService에서 난이도(difficulty)에 따라 부여
+        score: grade.score,
         feedback: grade.feedback,
       };
     });
   }
 
-  private getGradingSchema() {
+  private getGradingSchema(questionType: 'short_answer' | 'essay') {
```
[the remark this answers is not in the document]
Comment (NAKTA-Y) [reply packages/backend/src/quiz/quiz.service.ts]: 넵 개인적으로는 isEssay로 판단하는 것보다 type으로 분류해서 case에 따라서 output을 따로 보내주는게 더 깔끔해보이네요
